### PR TITLE
fix(inference): unify native weight ingress validation

### LINKED ISSUE #736: bug(inference): Q4/F16 direct-load path never validates tensor geometry against the model config
**Impact:** High — malformed, truncated, or transposed weight artifacts that are internally self-consistent load successfully and either produce out-of-bounds reads or silently wrong inference output, with no error surfaced.

**Location:** `crates/inference/src/forward/metal_qwen35.rs`. The Q4 weight loader validates only that a `.q4` file's on-disk header is self-consistent with its own payload length; it never compares the tensor's declared/actual dimensions against the model's configuration (hidden size, vocab size, intermediate size, head dims, etc). All per-layer attention/FFN projections and the embedding/`lm_head` tensors go through this path. The only geometry assertions present (`debug_assert_eq!` on FFN gate/up projection lengths) compile out entirely in release builds. Downstream, the public per-token forward function's only bounds check on the input token ID is against the configured vocabulary size, not against the actual size of the loaded embedding buffer — so a token ID that is legitimately in range per config can still read past the end of an undersized embedding buffer that was itself accepted by the loader.

**Repro sketch:**
1. Produce a Q4-format checkpoint directory where the embedding tensor has an internally valid Q4 header (self-consistent payload bounds) but fewer rows than the model config's vocabulary size, or where a projection tensor is transposed to the same total element count as expected.
2. Load the checkpoint via the direct Q4-directory constructor: no error is raised at load time.
3. Run a forward pass with a token ID that is valid per the model config's vocabulary size but exceeds the actual (undersized) embedding buffer's row count: the row read walks past the end of the allocation.
4. Alternatively, a same-element-count transposed projection loads successfully and produces silently incorrect (not just out-of-bounds) logits — no failure signal at all.

**Suggested fix:** Give every tensor loader an expected shape derived from the model configuration and return an error on any mismatch, using checked (not narrowing/unchecked) arithmetic for shape products. Apply this uniformly to Q4 and F16 artifacts (the F16 half is #735). Replace the debug-only FFN length assertions with a release-active `Result`-returning check. Derive the embedding-buffer bound actually used at inference time from the loaded artifact's real size (or validate it equals the configured vocabulary size at load time), not solely from the static config value.

Related: #655 (centralizing Q4/QuaRot manifest parsing) is the natural home for this validation once it exists — but the missing geometry check is a defect independent of that refactor.

### LINKED ISSUE #802: bug(inference): unify native Q4/KHF1 file validation — non-finite scale/bias, exact file extent, and panic-to-Result conversion
> _Issue authored by Claude (Anthropic agent) on behalf of @ohdearquant._

## What

The native `.q4` and KHF1 `.f16` file formats have validation gaps that are distinct from, and not covered by, the safetensors non-finite class tracked in issue #793:

- The CPU `.q4` loader (`load_q4_file`, `crates/inference/src/weights/q4_weights.rs`, roughly lines 808-879, verified in this checkout) reads the raw block payload via bounded `read_exact` (so truncation fails closed) and checks `shape.product() == original_len`, but **does not check whether the decoded per-block `scale`/`bias` f16 values are finite**, and **does not reject trailing bytes** left in the file after the expected block payload has been read.
- The Metal no-copy mmap path (`mmap_q4_weight` in `crates/inference/src/forward/metal_qwen35.rs`, roughly lines 774-811, calling `validate_q4_header_payload_bounds`) validates the header and payload bounds before handing the mapping to the GPU, but shares the same two gaps: no scale/bias finiteness check, and trailing bytes past the payload are accepted rather than rejected.
- The KHF1 `.f16` loader (`load_f16_tensor_file`, same file, roughly lines 895-945) fails closed on magic/version and uses `read_exact`, but **silently accepts** a header whose `shape.product() != numel` (unlike the sibling `.q4` loader, which does check this), and, like the `.q4` loaders, accepts NaN/Inf decoded values and trailing bytes.
- Separately, the Q4 source-side quantizers (`quantize_row_q4_0`, `quantize_tensor_q4_0`, `quantize_bf16_to_q4`, `quantize_f32_to_q4`, `quantize_f64_to_q4[_mode]`, roughly lines 333-570) already reject non-finite *source* values (verified: `quantize_block_with_mode_len` at line ~259 rejects any non-finite input element before deriving a block), but there is no check on the *output* side — a finite source value whose derived scale or bias overflows the stored f16 representation (i.e. encodes to infinity) is written to the block without error. Several of these public entry points also panic on shape mismatch rather than returning the `Result` they already declare (referenced around line ~419-432 of the design's source inspection; confirm exact line at implementation time as this file is actively changing).

## Why

A serialized Q4 or KHF1 file with a corrupted or maliciously crafted scale/bias, or with extra trailing bytes, currently loads successfully on both the CPU and Metal paths and produces silently wrong dequantized weights (NaN/Inf propagate into every matmul that touches that tensor) rather than failing at load time with a clear diagnostic. This is the same failure shape as the safetensors non-finite class in #793, but in the native on-disk format that #793 does not touch, and it additionally exposes format-integrity gaps (trailing bytes, KHF1 shape/numel mismatch) that have no safetensors analog.

**Exact-extent policy (binding maintainer decision):** rejecting trailing bytes after the expected payload is the default behavior this issue must implement. Before this default lands, verify against at least one real, currently-deployed `.q4` artifact (not just synthetic test fixtures) that no legitimate producer intentionally appends trailing data (e.g. a checksum, extension block, or padding) after the payload. If a real artifact is found with trailing bytes that are expected/benign, that changes this issue's default from reject-by-default to a documented allowance, and should be raised back to the maintainer before merge rather than silently choosing either behavior.

## Steps

1. Refactor a single, CPU-only, GPU-free "validated Q4 file" helper (operating on either a `File` or an in-memory byte slice) that both `load_q4_file` (CPU) and `mmap_q4_weight`'s preflight (Metal) call, so the same checks run on both paths and Metal validation is testable without a GPU.
2. In that helper, add: (a) a check that the file/mapping contains exactly the expected number of bytes for its declared shape and block count — no fewer (truncation) and, per the default above, no more (trailing bytes) — and (b) a check that every block's decoded `scale` and `bias` are finite, with `scale` additionally required to be strictly positive (a zero or negative scale is not a valid dequantization factor).
3. Apply the equivalent shape-product-vs-numel check already present in the CPU `.q4` loader to the KHF1 `.f16` loader (`load_f16_tensor_file`), so a KHF1 header whose shape disagrees with its element count fails closed instead of silently accepted.
4. On the quantizer side, add a serialized-metadata postcondition: after deriving a block's `scale`/`bias` from a finite source value, check that the derived f16 representation is itself finite before writing the block; if it overflows (i.e. the source value's magnitude exceeds what f16 scale/bias can represent), return `Err` rather than silently serializing an infinity. Do not clamp to max-finite — clamping would silently and incorrectly rescale the tensor's dequantized values.
5. Convert the identified panic-on-shape-mismatch call sites in the Q4 construction/quantization functions to return their already-declared `Result` error type instead of panicking. Re-verify the exact line numbers in the current source before making this change, since this file is under active development in this cluster; if a given panic site's function signature does not currently return `Result` at all, treat that specific conversion as **out of scope for this issue** and flag it separately, since changing a panic-only public function to return `Result` is a **major** semver event per the design's constraint, not a patch.

## Gates

- `cargo test -p lattice-inference` and full workspace `cargo test` pass, including new fixtures: a hand-authored `.q4` file with a NaN scale, an infinite scale, a zero scale, and a NaN bias, each rejected by both the CPU validated-file helper and the GPU-free Metal mmap preflight; a `.q4` file with one byte removed from the end (truncation) and one extra byte appended (trailing-bytes) both rejected; a KHF1 `.f16` file whose header `shape.product() != numel` rejected; a quantizer fixture with the largest finite source value that overflows f16 scale/bias metadata returning `Err` with no output file written.
- Mutation-sensitivity: each fixture above must be shown to fail when its corresponding check is reverted (per this project's mutation-sensitive-test convention, following the pattern at `q4_weights.rs` around lines 2494-2520).
- **Real-artifact verification gate (must be satisfied before merge, not just before this issue is opened):** run the trailing-bytes rejection check against at least one real, currently-deployed `.q4` artifact produced by this repository's own `quantize_q4` tool (and, if available, any externally-distributed `.q4` artifact already in use) to confirm no legitimate producer appends trailing data. Record the artifact(s) checked and the result in the PR description. If a real producer is found to intentionally append trailing bytes, stop and raise the discrepancy for a maintainer decision before changing the default to reject.
- Panic-to-`Result` conversions: any call site converted must be covered by a test that previously would have panicked and now returns a typed error; do not merge a conversion without a corresponding regression test.
- No cross-endian format change: this issue explicitly does not touch the Q4/KHF1 formats' current native-endian (`u16::from_ne_bytes`) read/write behavior — that is separate, out-of-scope debt (see the cluster index).
- Semver class: **patch** for loader/quantizer functions whose signatures already return `Result` (previously-accepted corrupt files newly fail to load; previously-silent overflow newly returns `Err`). Any panic-only public function whose signature would need to change to `Result` as part of the panic-to-`Result` conversion is **major** and is explicitly **not approved** as part of this issue — carve it out into a separate, clearly-labeled breaking-change proposal if found.

## Sequencing

- **Depends on:** the ingress-validation-seam issue (shares the same `IngestedTensor`/`validate_ingested_tensor` contract for the Q4/KHF1 payload variants) and the half-bits consolidation issue (scale/bias finiteness checks decode through the shared `half_bits` module).
- **Depended on by:** none directly in this cluster.
- **Related, not duplicated:** issue #736 (`Q4/F16 direct-load path never validates tensor geometry against the model config`) targets a distinct cross-check — whether the loaded tensor's shape agrees with the *model configuration* it's being loaded into, not whether the file's own internal shape/numel/scale/bias metadata is self-consistent. Issue #735 (`F16 tensor loader does not verify declared shape matches declared element count`) overlaps most closely with this issue's KHF1 shape/numel fix (step 3) — the two should be reconciled at implementation time so the same check isn't authored twice; whichever PR lands first should close or update the other. Issue #793 (`f32/f16 weight loaders accept non-finite values`) is the safetensors-side sibling of this issue's native-format non-finite checks and should be referenced, not duplicated.

### LINKED ISSUE #985: bug(inference): read_q4_header cursor-position doc-contract is false (BufReader advances fd past payload_offset)
## Summary

`read_q4_header` (`crates/inference/src/weights/q4_weights.rs:606`) documents "On return the file cursor is positioned at the start of the block data," but it reads the header through `std::io::BufReader::new(file)` over a `&File`. `BufReader` fills in ~8KB chunks, so for any header smaller than the buffer it advances the underlying shared file-descriptor cursor **past** `payload_offset` (it reads part of the block payload into the buffer). After the function returns and the `BufReader` is dropped, the raw `File` cursor is left past the payload start, so the documented post-condition is false.

This is the exact same defect fixed for `read_q3_header` in #978 — surfaced by that PR's sibling-path grep.

## Severity: latent (contract-only, no live misread today)

Every current caller either seeks explicitly before reading the payload or mmaps the whole file and indexes by `payload_offset`, so none relies on the returned cursor:

- `read_q4_payload_bounded` — calls `f.seek(SeekFrom::Start(header.payload_offset))` before `read_to_end`
- the Metal no-copy loaders in `forward::metal_qwen35` and `forward::moe_expert_cache` — mmap the file and index by `payload_offset`
- the CLI header-only path in `bin/lattice.rs` — reads no blocks

So q4 loading is correct in production. The risk is a future caller that trusts the doc-promised cursor and reads blocks off the same handle — it would silently read wrong bytes.

## Fix (mirror #978)

Change `read_q4_header` to take `&mut std::fs::File`, scope the `BufReader` so it drops before seeking, then `file.seek(SeekFrom::Start(payload_offset))?` so the documented cursor position actually holds. Update the small number of call sites to pass `&mut file`. Add a regression test that reads the first block directly off the same handle after `read_q4_header` and asserts it equals the first written block (mutation-sensitive: fails if the seek is reverted).

This could alternatively be folded into the shared Q4 ingress/validation refactor tracked in #801/#802.

## Provenance

Found by the sibling-invocation-path grep mandated after the #978 `read_q3_header` fix.

### PR DESCRIPTION
**Pull request by ohdearquant**
> _PR description authored by Codex (OpenAI agent) on behalf of @ohdearquant._

Closes #802. Closes #736. Closes #985.
Routes native Q4/KHF1 files through shared ingress validation, rejecting invalid metadata, inconsistent geometry, and non-exact payload extents before use.
Validates direct-load shapes against model config and restores the read_q4_header cursor contract; a quantize_q4-produced Qwen3.5 0.8B embedding artifact had zero trailing bytes.

## bench-compare disposition

Measured in a quiet window at the head named below. **The head has since advanced by four
commits** (a merge of `main`, the degenerate-scale fix, a formatting change, and a GDN norm
shape fix), so the figures in this table were not re-measured at the current head. What has
been re-established at the current head is the structural claim, in the second table below;
the magnitude was never certified, as stated at the end of this section.

```
base                7ba6f096eeba8e9bba8b195bd6d8f26e7ceb27f7
head                3a2d4e4af550f6faf6864b5603ce013c812202fb
resolution          --quick
targets             lattice-inference:elementwise_cpu_bench, lattice-embed:simd
inference features  <none>
enforcement         report-only
idle samples        before base 70.3% | between phases 75.8% | after head 78.3%  (floor 70.0%)
```

Gate output: 1 FAIL, 1 confirmed improvement, 246 informational (quick-mode SIMD micro-benches below resolution, #714).

| Bench | Δ point | 95% CI | new ns | base ns |
|---|---:|---|---:|---:|
| `residual_add_layer_norm/unfused/hidden768_seq128` | +9.11% | [+7.45%, +10.82%] | 100392.4 | 92013.9 |
| `layer_norm/4096` | −6.90% | [−8.92%, −4.83%] | 970.1 | 1042.0 |

**Neither figure is attributable to this change.** Both benched functions are defined in
`crates/inference/src/forward/cpu/norm.rs`, which this PR does not modify. Established by blob
identity across the two commits rather than by absence from a diff listing, with a known-changed
file as the control:

| file | base | head | |
|---|---|---|---|
| `crates/inference/src/forward/cpu/norm.rs` | `89115f15` | `89115f15` | identical |
| `crates/inference/benches/elementwise_cpu_bench.rs` | `d5730730` | `d5730730` | identical |
| `crates/inference/src/weights/ingress.rs` (control) | `2d51a2ab` | `9b234d3d` | differs |

The benched source and the bench harness are byte-identical between the two arms, so the diff
cannot reach these kernels semantically.

Re-verified at the **current** head `96cb4bd4d5754de639205ab9e1ea9a554fd87379` against current
`main` `5d54110835e9ef8d8a2b9fbef193cd46b6b7b75e`, same method, same control:

| file | base | head | |
|---|---|---|---|
| `crates/inference/src/forward/cpu/norm.rs` | `89115f15` | `89115f15` | identical |
| `crates/inference/benches/elementwise_cpu_bench.rs` | `d5730730` | `d5730730` | identical |
| `crates/inference/src/weights/ingress.rs` (control) | `2d51a2ab` | `73fcac84` | differs |

The precise scope of that claim: it rules out a semantic regression, not a timing difference.
The changed files live in the same crate the bench links, so codegen-unit partitioning and
inlining decisions can still move the machine code generated from identical source. This is
therefore *not* the compiled-out case in which base and head bench binaries have identical
effective source — `weights/q4_weights.rs` and the other changed files are compiled in.

What is consequently **not** certified is the magnitude. No A/A control was obtained in this
session, so there is no measured noise floor for this machine at this hour to judge +9.11%
against. A previously recorded A/A run of this harness on identical source produced a +8.95%
single-group FAIL (#1137); that run was on different hardware and so does not establish this
machine's floor, but it does establish that this harness can report a FAIL of the observed
magnitude with no code change at all.

No action taken on the FAIL, and it is not being reported as a regression. If a calibrated
number is wanted before merge, it needs an A/A control in the same window followed by a
`--full` re-run of the failing group.

### CODE REVIEW

**Review by ohdearquant — COMMENTED** [review]
Reviewed and approved at 96cb4bd4: per-head RMSNorm expectation for linear_attn.norm.weight now matches the safetensors loader and the checkpoint's own tensor geometry; native weight ingress validation unified across loaders with shape expectations stated at each call site. All five workflows green at this head, including the Q4 composed golden and perplexity regression gates. (Posted as comment: repository single-account topology blocks formal self-approval; the branch ruleset requires 0 approvals, so this comment is the review of record.)
